Intent Support

Brayzure opened the pull request:
Initial PR
- Adds support for registering intents.

ToDo
- Check to make sure this doesn't horrifically break things, and fix conflicts as they arise.

Also worth considering, should intent defining be moved elsewhere? It's going to be required in gateway v7 (whenever it releases), so moving it out of options into its own argument preemptively may be a good idea.

Brayzure committed:
Preliminary Intent Support
- Adds support for registering intents
- Cache not being guaranteed not addressed yet

Brayzure committed:
Linting after the fact. Ooops

Brayzure committed:
Remove minor support for edge case

Brayzure committed:
Merge branch 'dev' of https://github.com/abalabahaha/eris into dev

Brayzure committed:
Typings

Brayzure committed:
More explicit typing

Brayzure committed:
Intent-related ws close codes

Brayzure committed:
Add intents example

Brayzure committed:
Merge branch 'dev' into dev

Astraeus- commented on lib/Client.js:
```diff
@@ -98,6 +98,7 @@ class Client extends EventEmitter {
     * @arg {Object} [options.agent] A HTTP Agent used to proxy requests
     * @arg {Number} [options.maxReconnectAttempts=Infinity] The maximum amount of times that the client is allowed to try to reconnect to Discord.
     * @arg {Function} [options.reconnectDelay] A function which returns how long the bot should wait until reconnecting to Discord.
+    * @arg {Number | String[]} [options.intents] A list of intents, or raw bitmask value describing the intents to subscribe to. "presence" intent must be enabled on your application's page to be used.
```
`Guild members` also needs to be enabled on the application's page to be used as per https://github.com/discordapp/discord-api-docs/issues/1363#issue-565525845

Brayzure committed:
Fix ws close code for invalid intents

Brayzure committed:
Fix guild member requests when using intents

Brayzure committed:
Add checks against requesting guild members without guildMembers intent

Brayzure committed:
If guildPresences intent isn't specified, get user updates from guildMemberUpdate instead

Brayzure committed:
Linting issue

Zoddo commented on lib/Constants.js:
```diff
@@ -201,3 +201,21 @@ module.exports.ChannelTypes = {
+    directMessages:         1 << 12,
+    directMessageReactions: 1 << 13,
+    directMessageTyping:    1 << 14
+};
```
Maybe we can use the same format than in the [Discord's documentation](https://discordapp.com/developers/docs/topics/gateway#list-of-intents) (ex: `GUILD_MEMBERS` instead of `guildMembers`), like what we did for OPCodes?

Brayzure replied:
Would prefer to keep it as is, since it's the same format as we use for event names. It's formatted like that in Constants because intent integer calculating checks directly against Constants.

Brayzure replied:
See also: Permissions, which use camelCase. There's nowhere in the lib where a dev using the lib has to use Discord case, would not be the greatest idea to change that now.

Brayzure committed:
Guild chunk request bugfix
- Throw an error if user attempts to request a guild's members multiple times before the first request finishes

Brayzure committed:
Add priority to bucket calls
- Buckets now allow you to reserve a certain number of tokens for priority calls. These tokens are not consumed unless a priority call is added to the queue

abalabahaha commented on lib/util/Bucket.js:
```diff
@@ -13,23 +13,30 @@ class Bucket {
     * @arg {Number} latencyRef.latency Interval between consuming tokens
     */
-    constructor(tokenLimit, interval, latencyRef = {latency: 0}) {
+    constructor(tokenLimit, interval, reserved = 0, latencyRef = {latency: 0}) {
```
Not a fan of adding it like this as it breaks things... Could use an options object as 4th param

Brayzure replied:
Strikes me as a bit odd to do it like that. Alternatively, we could merge `latencyRef` and `reserved` into a single options object. I don't even see anywhere in the lib where we explicitly define `latencyRef`, though search could be failing me.

apacheli commented on lib/Client.js:
```diff
@@ -147,6 +148,24 @@ class Client extends EventEmitter {
             this.options.ws.agent = this.options.agent;
         }
 
+        if(this.options.intents) {
```
This will cause unintended behavior if someone sets `intents: 0` and `getAllUsers: true`

Brayzure replied:
Ah yes, took me a moment to realize what you meant. Well, checking for the existence of the property instead should work. Thanks for pointing that out.

Brayzure committed:
Update Bucket constructor and fix intents bug
- Bucket constructor now takes options object, replaces latencyRef parameter and previous reserved parameter
- Fix case where some intent-specific logic wouldn't run if the intents value was specified to be 0

Brayzure committed:
Add missing typing
